records/2021-07-16.scalability.database-bottleneck.md

```diff
@@ -0,0 +1,135 @@
+to appear at that point.
+12. The API instances are down as these are detached from the load balancers,
+the database is overloaded with connections and can't take any more. **We are 
+losing notifications** because we don't save these when the API is unavailable.
+
+[1] The ALB makes health checks on Notify's stack components to make sure these
+are eligible for traffic, but they call a `_status?simple=true` endpoint that makes
+sure a database call is not included as part of the check. This means the situation
+might be more dire as the instance overall slowed down so much due to the database
+might lock the process due to its unavailability.
+
+_TODO: Add sequence or flow chart_
+
+## Options
+
+### Stream the notifications save
+
+Upon receiving a notification from the API, do not save it in the current
+*notification-api* process but defer it to a celery task.
+
+### Insert the notifications as a batch
+
+As we streamline the notification save with the previous option and got
+these under Celery land, we can now batch these easily. There is a 
+[celery-batch](https://pypi.org/project/celery-batches/) project which 
+exists that might offer us the option to easily achieve this with minimal
+setup.
+
+### Tweaking autovacuum options
+
+There are a few autovacuum configuration that we can tweak in order to delay
+the autovacuum action that triggers a performance slowdown of our database:
+
+* **autovacuum_scale_factor** (default: 0.2): we could use this to delay the vacuum, (suggestion 0.3).
+* **cost_delay**: We could use this to delay by increasing the threshold.
+
+### Manual execution of the vacuum process on the `notifications` table
+
+An option is to keep the autovacuum enabled but trigger a manual vacuum prior
+to traffic time. This would better prepare the database for upcoming load.
+
+Don't run `vacuum FULL`: this will take on most database resources and prevent
+proper operation of the database.
+
+### Disable the autovacuum process on the `notifications` table
+
+To avoid the autovacuum to process the `notifications` table during the daily
+blast of notifications, we could disable the autovacuum to prevent it from
+triggering. This need to be done carefully, as a maximum limit of 4 billion
+transactions can occur in PostgreSQL. The autovacuum cleanup reset this
+transaction number. If this threshold is reached, our database will be severely
+affected.
+
+This is a risky move and certainly not a long/mid term solution.
+
+### Database connections pooling
+
+To maximise the distribution of database connection, we can have a
+pooler that would sit in-between the Notify stack and the AWS RDS
+Aurora Postgres instances.
+
+RDS Proxies would exists for this purpose. Apparently, we could create an
+endpoint each for both readers and the writer. This could co-exist nicely
+with our existing application logic which dispatch database sessions to
+either the readers or writer.
```

Comment:
This seems like it should be the easiest solution and **should** have no impact on your application.

Reply:
Yep; we should do it no matter what.